refactor: Enforce C-str fmt strings in WalletLogPrintf()

maflcko opened the pull request:
All fmt functions only accept a raw C-string as argument.

There should never be a need to pass a format string that is not a compile-time string literal, so disallow it in `WalletLogPrintf()` to avoid accidentally introducing it.

Apart from consistency, this also fixes the clang-tidy plugin bug https://github.com/bitcoin/bitcoin/pull/26296#discussion_r1286821141.

Commit:
doc: Fix bitcoin-unterminated-logprintf tidy comments

* Move module description from test to LogPrintfCheck
* Add test doc
* Remove unused comment, see https://github.com/bitcoin/bitcoin/pull/26296/files#r1279351539

Commit:
refactor: Enforce C-str fmt strings in WalletLogPrintf()

theuni commented on src/wallet/wallet.cpp:
```diff
@@ -2319,7 +2319,7 @@ OutputType CWallet::TransactionChangeType(const std::optional<OutputType>& chang
 {
     LOCK(cs_wallet);
-    WalletLogPrintf("CommitTransaction:\n%s", tx->ToString());
+    WalletLogPrintf("CommitTransaction:\n%s", tx->ToString()); // NOLINT(bitcoin-unterminated-logprintf)
```
Hmm, this one wasn't being caught before?

theuni replied:
Hmm.

I haven't tried this using the c-i scripts, but running clang-tidy manually against current master, this works fine for me as-is:
```bash
$ /opt/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-tidy-16 --load=/home/cory/dev/bitcoin2/contrib/devtools/bitcoin-tidy/build2/libbitcoin-tidy.so -checks="-*,bitcoin-*" wallet/wallet.cpp -- -DHAVE_BUILD_INFO -I. -I./minisketch/include -I./secp256k1/include -I./univalue/include -I./leveldb/include -isystem /home/cory/dev/bitcoin2/depends/x86_64-pc-linux-gnu/include -DBOOST_MULTI_INDEX_DISABLE_SERIALIZATION -DBOOST_NO_CXX98_FUNCTION_BASE -DHAVE_CONFIG_H  -std=c++17
1 warning generated.
/home/cory/dev/bitcoin2/src/wallet/wallet.cpp:2322:21: error: Unterminated format string used with LogPrintf [bitcoin-unterminated-logprintf,-warnings-as-errors]
    WalletLogPrintf("CommitTransaction:\n%s", tx->ToString());
                    ^
                                           \n
1 warning treated as error
```

Any idea what the difference is?

theuni replied:
Huh. For whatever reason, my clang-tidy-16 acts as if `IgnoreUnlessSpelledInSource` is set. That's weird.

It came from here: https://github.com/llvm/llvm-project/releases/download/llvmorg-16.0.0/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz

I can't reproduce with clang-tidy-15. clang-query-16 only matches if traversal is set to `IgnoreUnlessSpelledInSource`.

> $ /opt/clang+llvm-15.0.2-x86_64-unknown-linux-gnu/bin/clang-query wallet/wallet.cpp -- -DHAVE_BUILD_INFO -I. -I./minisketch/include -I./secp256k1/include -I./univalue/include -I./leveldb/include -isystem /home/cory/dev/bitcoin2/depends/x86_64-pc-linux-gnu/include -DBOOST_MULTI_INDEX_DISABLE_SERIALIZATION -DBOOST_NO_CXX98_FUNCTION_BASE -DHAVE_CONFIG_H  -std=c++17 -O2
> clang-query> m cxxMemberCallExpr(thisPointerType(qualType(hasDeclaration(cxxRecordDecl(hasName("CWallet"))))),callee(cxxMethodDecl(hasName("WalletLogPrintf"))),hasArgument(0, stringLiteral().bind("logstring")))
> 0 matches.
> clang-query> set traversal IgnoreUnlessSpelledInSource
> clang-query> m cxxMemberCallExpr(thisPointerType(qualType(hasDeclaration(cxxRecordDecl(hasName("CWallet"))))),callee(cxxMethodDecl(hasName("WalletLogPrintf"))),hasArgument(0, stringLiteral().bind("logstring")))
> 
> ...
> 
> Match #33:
> 
> /home/cory/dev/bitcoin2/src/wallet/wallet.cpp:4149:36: note: "logstring" binds here
>             wallet.WalletLogPrintf("Making a new watchonly wallet containing the unwatched solvable scripts\n");
>                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/cory/dev/bitcoin2/src/wallet/wallet.cpp:4149:13: note: "root" binds here
>             wallet.WalletLogPrintf("Making a new watchonly wallet containing the unwatched solvable scripts\n");
>             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 33 matches
> 
> 


I don't think this is worth spending any more time on, but I still don't understand why it's different.

theuni commented on src/wallet/wallet.h:
```diff
@@ -890,9 +890,10 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
-    void WalletLogPrintf(std::string fmt, Params... parameters) const {
-        LogPrintf(("%s " + fmt).c_str(), GetDisplayName(), parameters...);
+    template <typename... Params>
+    void WalletLogPrintf(const char* fmt, Params... parameters) const
```
I was going to suggest the `nonnull` attribute here, but I believe our tidy check actually enforces that :)

theuni reviewed (commented):
Makes sense to me. I almost did the same thing while poking at `WalletLogPrintf` for the sake of easier static analysis.

Am I missing some obvious reason though, other than "might as well"?

Commit:
refactor: Enable all clang-tidy plugin bitcoin tests

This makes it easier to add new ones without having to modify this file
every time.

Commit:
bitcoin-tidy: Apply bitcoin-unterminated-logprintf to spkm as well

theuni commented:
ACK fa60fa3b0cba4a30726af8e0e9d1e84e14849eda
